File: src/main/python/AoC2015_13.py
#! /usr/bin/env python3
#
# Advent of Code 2015 Day 13
#

from __future__ import annotations

import itertools
import sys
from collections import defaultdict
from dataclasses import dataclass

from aoc.common import InputData
from aoc.common import SolutionBase
from aoc.common import aoc_samples

TEST = """\
Alice would gain 54 happiness units by sitting next to Bob.
Alice would lose 79 happiness units by sitting next to Carol.
Alice would lose 2 happiness units by sitting next to David.
Bob would gain 83 happiness units by sitting next to Alice.
Bob would lose 7 happiness units by sitting next to Carol.
Bob would lose 63 happiness units by sitting next to David.
Carol would lose 62 happiness units by sitting next to Alice.
Carol would gain 60 happiness units by sitting next to Bob.
Carol would gain 55 happiness units by sitting next to David.
David would gain 46 happiness units by sitting next to Alice.
David would lose 7 happiness units by sitting next to Bob.
David would gain 41 happiness units by sitting next to Carol.
"""


@dataclass
class Happiness:
    matrix: list[list[int]]

    @classmethod
    def from_input(cls, inputs: InputData) -> Happiness:
        cntr = itertools.count()
        idxs = defaultdict[str, int](lambda: next(cntr))
        values = dict[tuple[int, int], int]()
        for line in inputs:
            splits = line[:-1].split()
            key = (idxs[splits[0]], idxs[splits[10]])
            value = int(splits[3]) * (1 if splits[2] == "gain" else -1)
            values[key] = value
        matrix = [[0] * (len(idxs) + 1) for _ in range(len(idxs) + 1)]
        for k, v in values.items():
            matrix[k[0]][k[1]] = v
        return Happiness(matrix)

    def solve(self, size: int) -> int:
        return max(
            sum(
                self.matrix[d1][d2] + self.matrix[d2][d1]
                for d1, d2 in itertools.zip_longest(p, p[1:], fillvalue=p[0])
            )
            for p in itertools.permutations(range(size), size)
        )

    def get_optimal_happiness_change_without_me(self) -> int:
        return self.solve(len(self.matrix) - 1)

    def get_optimal_happiness_change_with_me(self) -> int:
        return self.solve(len(self.matrix))


Input = Happiness
Output1 = int
Output2 = int


class Solution(SolutionBase[Input, Output1, Output2]):
    def parse_input(self, input_data: InputData) -> Input:
        return Happiness.from_input(input_data)

    def part_1(self, happiness: Happiness) -> Output1:
        return happiness.get_optimal_happiness_change_without_me()

    def part_2(self, happiness: Happiness) -> Output2:
        return happiness.get_optimal_happiness_change_with_me()

    @aoc_samples((("part_1", TEST, 330),))
    def samples(self) -> None:
        pass


solution = Solution(2015, 13)


def main() -> None:
    solution.run(sys.argv)


if __name__ == "__main__":
    main()

File: src/main/java/AoC2015_13.java
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.IntStream;

import com.github.pareronia.aoc.IterTools;
import com.github.pareronia.aoc.MutableInt;
import com.github.pareronia.aoc.solution.Sample;
import com.github.pareronia.aoc.solution.Samples;
import com.github.pareronia.aoc.solution.SolutionBase;

public class AoC2015_13 extends SolutionBase<AoC2015_13.Happiness, Integer, Integer> {
    
    private AoC2015_13(final boolean debug) {
        super(debug);
    }

    public static final AoC2015_13 create() {
        return new AoC2015_13(false);
    }

    public static final AoC2015_13 createDebug() {
        return new AoC2015_13(true);
    }
    
    @Override
    protected Happiness parseInput(final List<String> inputs) {
        return Happiness.fromInput(inputs);
    }

    @Override
    public Integer solvePart1(final Happiness input) {
        return input.getOptimalHappinessChangeWithoutMe();
    }
    
    @Override
    public Integer solvePart2(final Happiness input) {
        return input.getOptimalHappinessChangeWithMe();
    }

    @Override
    @Samples({
        @Sample(method = "part1", input = TEST, expected = "330"),
    })
    public void samples() {
    }

    public static void main(final String[] args) throws Exception {
        AoC2015_13.create().run();
    }
    
    private static final String TEST =
            """
    	Alice would gain 54 happiness units by sitting next to Bob.\r
    	Alice would lose 79 happiness units by sitting next to Carol.\r
    	Alice would lose 2 happiness units by sitting next to David.\r
    	Bob would gain 83 happiness units by sitting next to Alice.\r
    	Bob would lose 7 happiness units by sitting next to Carol.\r
    	Bob would lose 63 happiness units by sitting next to David.\r
    	Carol would lose 62 happiness units by sitting next to Alice.\r
    	Carol would gain 60 happiness units by sitting next to Bob.\r
    	Carol would gain 55 happiness units by sitting next to David.\r
    	David would gain 46 happiness units by sitting next to Alice.\r
    	David would lose 7 happiness units by sitting next to Bob.\r
    	David would gain 41 happiness units by sitting next to Carol.""";
    
    record Happiness(int[][] happinessMatrix) {
        
        public static Happiness fromInput(final List<String> inputs) {
            final Map<String, Integer> map = new HashMap<>();
            final Map<int[], Integer> values = new HashMap<>();
            final MutableInt cnt = new MutableInt(0);
            for (final String input : inputs) {
                final String[] s = input.substring(0, input.length() - 1).split(" ");
                final String d1 = s[0];
                final String d2 = s[10];
                final int idx1 = map.computeIfAbsent(d1, x -> cnt.getAndIncrement());
                final int idx2 = map.computeIfAbsent(d2, x -> cnt.getAndIncrement());
                final int value = Integer.parseInt(s[3]);
                values.put(new int[] { idx1, idx2 }, "gain".equals(s[2]) ? value : -value);
            }
            final int[][] happinessMatrix = new int[map.size() + 1][map.size() + 1];
            values.entrySet().stream()
                .forEach(e -> {
                    happinessMatrix[e.getKey()[0]][e.getKey()[1]] = e.getValue();
                });
            return new Happiness(happinessMatrix);
        }
        
        private int solve(final int size) {
            final int[] idxs = IntStream.range(0, size).toArray();
            return IterTools.permutations(idxs)
                .mapToInt(p ->
                    IntStream.range(0, p.length)
                        .map(i -> {
                            final int d1 = p[i];
                            final int d2 = p[(i + 1) % p.length];
                            return this.happinessMatrix[d1][d2] + this.happinessMatrix[d2][d1];
                        })
                        .sum())
                .max().orElseThrow();
        }
        
        public int getOptimalHappinessChangeWithoutMe() {
            return solve(this.happinessMatrix.length - 1);
        }
        
        public int getOptimalHappinessChangeWithMe() {
            return solve(this.happinessMatrix.length);
        }
    }

}